### PR TITLE
Add extension hooks for encoding and decoding UDAFs and UDWFs

### LINKED ISSUE #11422: Handle Serde for Custom AggregateUDFImpl traits
### Is your feature request related to a problem or challenge?

Currently, `PhysicalExtensionCodec` and `LogicalExtensionCodec` provide hooks for decoding and encoding custom scalar UDFs. But we're missing the same for aggregation UDFs.

### Describe the solution you'd like

Add `try_encode_udaf` and `try_decode_udaf` functions to `PhysicalExtensionCodec` and `LogicalExtensionCodec` with default noop implementations. Add the required function definition fields (optional bytes) to the Proto messages.

### Describe alternatives you've considered

There is no alternative currently. It's not possible to run a parameterized aggregation UDF in a distributed architecture.

### Additional context

See https://github.com/apache/datafusion/issues/8706

### PR DESCRIPTION
**Pull request by joroKr21**
## Which issue does this PR close?

Closes #11422.

## Rationale for this change

The rationale for this change is to be able to run parameterized aggregation UDFs in a distributed architecture.
E.g. one use case is implementing an `sample` UDF (akin to `array_agg` but with a limit on the number of elements collected).

## What changes are included in this PR?

* Extend the interface of `LogicalExtensionCodec` to handle aggregation UDFs and window UDFs.
* Extend the interface of `PhysicalExtensionCodec` to handle aggregation UDFs.
    * I couldn't figure out physical window UDFs so those are omitted.
* Handle `ignore_nulls` and `distinct` flags better.

## Are these changes tested?

Yes, added roundtrip tests for encoding and decoding.

## Are there any user-facing changes?

Yes, users can now override the new encoding and decoding functions.

### CODE REVIEW

**Review by avantgardnerio — APPROVED** [review]
LGTM

**Review by thinkharderdev — APPROVED** [review]
nice!

File: datafusion-examples/examples/composed_extension_codec.rs [hunk 1]
```diff
@@ -239,53 +240,52 @@ struct ComposedPhysicalExtensionCodec {
     codecs: Vec<Arc<dyn PhysicalExtensionCodec>>,
 }
 
+impl ComposedPhysicalExtensionCodec {
```
Comment (alamb) [inline datafusion-examples/examples/composed_extension_codec.rs:243]: FYI @lewiszlw

**Comment by alamb** [comment]
Thanks for the reviews @thinkharderdev  and @avantgardnerio  -- @joroKr21  it appears this branch has conflicts that need to be resolved. Once those are sorted out I think we can merge this PR

File: datafusion/proto/src/physical_plan/mod.rs [hunk 2]
```diff
@@ -2018,6 +2021,16 @@ pub trait PhysicalExtensionCodec: Debug + Send + Sync {
     fn try_encode_udf(&self, _node: &ScalarUDF, _buf: &mut Vec<u8>) -> Result<()> {
         Ok(())
     }
+
+    fn try_decode_udaf(&self, name: &str, _buf: &[u8]) -> Result<Arc<AggregateUDF>> {
+        not_impl_err!(
+            "LogicalExtensionCodec is not provided for aggregate function {name}"
```
Comment (lewiszlw) [inline datafusion/proto/src/physical_plan/mod.rs:2027]:
```suggestion
            "PhysicalExtensionCodec is not provided for aggregate function {name}"
```

File: datafusion-examples/examples/composed_extension_codec.rs (hunk 1 above)
Comment (lewiszlw) [reply datafusion-examples/examples/composed_extension_codec.rs:243]: 👍

**Comment by joroKr21** [comment]
@alamb do you get this issue locally on `main`?
```
error[E0432]: unresolved import `chrono::MappedLocalTime`
  --> datafusion/functions/src/datetime/to_local_time.rs:34:24
   |
34 | use chrono::{DateTime, MappedLocalTime, Offset, TimeDelta, TimeZone, Utc};
   |                        ^^^^^^^^^^^^^^^ no `MappedLocalTime` in the root
```

**Comment by lewiszlw** [comment]
I got this error before, running `cargo update` fixed it.

**Comment by alamb** [comment]
> I got this error before, running `cargo update` fixed it.

🤔  looks like we probably need to increase the minimum version of chrono required to one that has `MappedLocalTime`

**Comment by alamb** [comment]
Thanks again @joroKr21  and @lewiszlw

**Pull request merged**